File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    Map<TreeNode, TreeNode> parent = new HashMap<>();
    Map<TreeNode, Integer> depth = new HashMap<>();
    int maxDepth = 0;

    private void dfs(TreeNode node, TreeNode par, int d) {
        if (node == null) return;

        parent.put(node, par);
        depth.put(node, d);
        maxDepth = Math.max(maxDepth, d);

        dfs(node.left, node, d + 1);
        dfs(node.right, node, d + 1);
    }

    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        dfs(root, null, 0);

        List<TreeNode> deepest = new ArrayList<>();
        for (TreeNode node : depth.keySet()) {
            if (depth.get(node) == maxDepth) {
                deepest.add(node);
            }
        }

        Set<TreeNode> curr = new HashSet<>(deepest);

        while (curr.size() > 1) {
            Set<TreeNode> next = new HashSet<>();
            for (TreeNode node : curr) {
                next.add(parent.get(node));
            }
            curr = next;
        }

        return curr.iterator().next();
    }
}

File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<TreeNode*, TreeNode*> parent;
    unordered_map<TreeNode*, int> depth;
    int maxDepth = 0;

    void dfs(TreeNode* node, TreeNode* par, int d) {
        if (!node) return;
        parent[node] = par;
        depth[node] = d;
        maxDepth = max(maxDepth, d);
        dfs(node->left, node, d + 1);
        dfs(node->right, node, d + 1);
    }

    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        dfs(root, nullptr, 0);

        vector<TreeNode*> deepest;
        for (auto &it : depth) {
            if (it.second == maxDepth)
                deepest.push_back(it.first);
        }

        unordered_set<TreeNode*> curr(deepest.begin(), deepest.end());

        while (curr.size() > 1) {
            unordered_set<TreeNode*> next;
            for (auto node : curr) {
                next.insert(parent[node]);
            }
            curr = next;
        }

        return *curr.begin();
    }
};